pkg/dss/scd/store.go

```diff
@@ -2,14 +2,34 @@ package scd
 
 import (
 	"context"
+	"time"
 
 	"github.com/golang/geo/s2"
 	dssmodels "github.com/interuss/dss/pkg/dss/models"
 	scdmodels "github.com/interuss/dss/pkg/dss/scd/models"
 )
 
-// Store abstracts interactions with a backing data store.
-type Store interface {
+// OperationStore abstracts operation-specific interactions with the backing data store.
+type OperationStore interface {
+	// GetOperation returns the operation identified by "id".
+	GetOperation(ctx context.Context, id scdmodels.ID) (*scdmodels.Operation, error)
+
+	// DeleteOperation deletes the operation identified by "id" and owned by "owner".
+	// Returns the deleted Operation and all Subscriptions affected by the delete.
+	DeleteOperation(ctx context.Context, id scdmodels.ID, owner dssmodels.Owner) (*scdmodels.Operation, []*scdmodels.Subscription, error)
+
+	// UpsertOperation inserts or updates an operation using key as a fencing
+	// token. If operation does not reference an existing subscription, an
+	// implicit subscription with parameters notifySubscriptionForConstraints
+	// and subscriptionBaseURL is created.
+	UpsertOperation(ctx context.Context, operation *scdmodels.Operation, key []scdmodels.OVN, notifySubscriptionForConstraints bool, subscriptionBaseURL string) (*scdmodels.Operation, []*scdmodels.Subscription, error)
```

Comment:
I don't think we want to put this much application logic in the storage layer.  I'd think the storage layer would just accept an Operation to upsert and wouldn't care about the OVNs (application layer would check the OVNs before calling the storage upsert).  Likewise, I would just have the application layer query for Subscriptions in the Operation's Volume4D (an existing storage method) rather than embedding that application logic in the storage layer.  I'd also have the application layer create the implicit Subscription.

Reply:
The problem then would be that the service logic needs to learn about transactions, and the storage layer would have to provide abstractions of those to ensure proper isolation of concurrent operations. That being said, we could probably move creation of implicit subscriptions to the service layer, but I would rather like to keep the check for fencing tokens in the storage layer.